64_MergeSort/main.cpp

```diff
@@ -0,0 +1,368 @@
+
+		for (size_t phaseIndex = 1; phaseIndex <= numberOfPhases; phaseIndex++)
+		{
+			smart_refctd_ptr<nbl::video::IGPUCommandBuffer> cmdbuf;
+			{
+				smart_refctd_ptr<nbl::video::IGPUCommandPool> cmdpool = m_device->createCommandPool(m_computeQueueFamily, IGPUCommandPool::CREATE_FLAGS::TRANSIENT_BIT);
+				if (!cmdpool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, 1u, &cmdbuf))
+				{
+					return logFail("Failed to create Command Buffers!\n");
+				}
+			}
+
+			cmdbuf->begin(IGPUCommandBuffer::USAGE::ONE_TIME_SUBMIT_BIT);
+			cmdbuf->beginDebugMarker("My Compute Dispatch", core::vectorSIMDf(0, 1, 0, 1));
+			cmdbuf->bindComputePipeline(m_computePipeline.get());
+
+			// If phase index is ODD, the inpt and output buffer are as usual. 
+			// If phase index is EVEN, input buffer is the output buffer of last phase.
+			if (phaseIndex % 2 == 1)
+			{
+				cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, m_pipelineLayout.get(), 0, 1, &m_descriptorSet1.get());
+			}
+			else
+			{
+				cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, m_pipelineLayout.get(), 0, 1, &m_descriptorSet2.get());
+			}
+
+			struct SortingPhaseData
+			{
+				uint32_t numElementsPerArray;
+				uint32_t bufferLength;
+			};
+
+			const SortingPhaseData pushConstantData = {
+				.numElementsPerArray = (uint32_t)pow(2, phaseIndex - 1),
+				.bufferLength = static_cast<uint32_t>(NumberOfElementsToSort),
+			};
+
+			cmdbuf->pushConstants(m_pipelineLayout.get(), nbl::asset::IShader::ESS_COMPUTE, 0u, sizeof(pushConstantData), &pushConstantData);
+
+			cmdbuf->dispatch((uint32_t)ceilf(NumberOfElementsToSort / powf(2.0f, phaseIndex)), 1, 1);
+			cmdbuf->endDebugMarker();
+
+			cmdbuf->end();
+
+			{
+				// queues are inherent parts of the device, ergo not refcounted (you refcount the device instead)
+				IQueue* queue = m_device->getQueue(m_computeQueueFamily, 0);
+
+				IQueue::SSubmitInfo submitInfos[1] = {};
+				// The IGPUCommandBuffer is the only object whose usage does not get automagically tracked internally, you're responsible for holding onto it as long as the GPU needs it.
+				// So this is why our commandbuffer, even though its transient lives in the scope equal or above the place where we wait for the submission to be signalled as complete.
+				const IQueue::SSubmitInfo::SCommandBufferInfo cmdbufs[] = { {.cmdbuf = cmdbuf.get()} };
+				submitInfos[0].commandBuffers = cmdbufs;
+
+				const IQueue::SSubmitInfo::SSemaphoreInfo signals[] = { {.semaphore = m_phaseSemaphore.get(),.value = ++monotonicallyIncreasingCounter,.stageMask = asset::PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT} };
+				submitInfos[0].signalSemaphores = signals;
+
+				queue->startCapture();
+				queue->submit(submitInfos);
+				queue->endCapture();
+
+			}
+			const ISemaphore::SWaitInfo wait_infos[] = { {
+		.semaphore = m_phaseSemaphore.get(),
+		.value = monotonicallyIncreasingCounter,
+	}, };
+
+			m_device->blockForSemaphores(wait_infos);
```

Comment:
use frames in flight properly, there's no need to make the CPU block... just create log2(N) commandbuffers you don't recycle

and use BDA, not descriptor sets then you don't need log2(N) descriptor sets

Reply:
actually, don't use log2(N) submits, use a single commandbuffer and dispatch

use pipeline barriers to synchronize between subsequent compute dispatches